Implement new Input Capture protocol

3l0w opened the pull request:
#### Describe your PR, what does it fix/add?

This PR implements the input capture protocol defined in hyprwm/hyprland-protocols#8.
It also fixed a bug where `wl_output::geometry` `x` and `y` arguments where always 0. 
The `wl_output::geometry` event is now sent when the output layout change (Help to follow `x` and `y` modifications).

#### Is it ready for merging, or does it need work?
It still need work, like sending the keyboard layout.

TODO List:
* [x] See if the fd multiplication problem is fixed
* [x] Calculate the invalids barriers on SetPointerBarriers
* [x] Replace the trash impl of get_line_intersection by one that follow the specs
* [x] Warp pointer when the cursor is released by the capturing program
* [x] Move most of the impl to Hyprland
* [x] Impl some kind of security around https://github.com/hyprwm/Hyprland/pull/9930, maybe move some of the impl ive done here to Hyprland
* [x] Fix the isLock shenanigans 
* [x] See if i limit to only one client, (for now multiple clients can connect, first arrive with his zones, first to be triggered)
* [x] Fix unable to access tty when captured

3l0w committed:
input-capture: impl protocol

Co-authored-by: Gerben Meijer <infernix@gmail.com>
Co-authored-by: Emerson Matos <emegson@live.com>
Co-authored-by: Mihai Fufezan <mihai@fufexan.net>

3l0w committed:
input-capture: impl protocol

Co-authored-by: Gerben Meijer <infernix@gmail.com>
Co-authored-by: Emerson Matos <emegson@live.com>
Co-authored-by: Mihai Fufezan <mihai@fufexan.net>
Co-authored-by: Ashley Stonham <reddrop@reddrop.net>
Co-authored-by: extra-salad <87635414+extra-salad@users.noreply.github.com>

Fix config

vaxerski commented on .gitmodules:
```diff
@@ -1,6 +1,7 @@
 	path = subprojects/hyprland-protocols
-	url = https://github.com/hyprwm/hyprland-protocols
+	url = https://github.com/3l0w/hyprland-protocols
+	branch = feat/input-capture-impl
```
marking this one as a blocker until this is removed so we dont forget (don't resolve)

vaxerski replied:
can be removed now

vaxerski commented on src/protocols/InputCapture.hpp:
```diff
@@ -0,0 +1,37 @@
+    void sendFrame();
+
+  private:
+    bool active;
```
initialize

vaxerski commented on src/protocols/core/Output.cpp:
```diff
@@ -55,6 +53,8 @@ void CWLOutputResource::updateState() {
     if (resource->version() >= 2)
         resource->sendScale(std::ceil(monitor->scale));
 
+    resource->sendGeometry(monitor->vecPosition.x, monitor->vecPosition.y, monitor->output->physicalSize.x, monitor->output->physicalSize.y, (wl_output_subpixel)monitor->output->subpixel, monitor->output->make.c_str(),
```
why this change?

3l0w replied:
This event should be sent when any of the properties change. But without this change this is not done.
In EIS or Input Capture portal, i need to notify whenever the output geometry change.
[wl_output::geometry](https://wayland.app/protocols/wayland#wl_output:event:geometry)

vaxerski replied:
ok

vaxerski commented on CMakeLists.txt:
```diff
@@ -332,6 +332,8 @@ protocolnew("staging/linux-drm-syncobj" "linux-drm-syncobj-v1" false)
 protocolnew("staging/xdg-dialog" "xdg-dialog-v1" false)
 protocolnew("staging/single-pixel-buffer" "single-pixel-buffer-v1" false)
 
+protocolnew("subprojects/hyprland-protocols/protocols" "hyprland-input-capture-v1"
```
why is there an empty line above?

there should be an empty line only to separate wayland from w-ps

```cpp
protocolnew(...)
protocolnew(...)
protocolnew(...)

protocolwayland()
```

vaxerski commented on src/protocols/InputCapture.cpp:
```diff
@@ -0,0 +1,75 @@
+void CInputCaptureProtocol::bindManager(wl_client* client, void* data, uint32_t ver, uint32_t id) {
+    const auto RESOURCE = m_vManagers.emplace_back(std::make_unique<CHyprlandInputCaptureManagerV1>(client, ver, id)).get();
+
+    RESOURCE->setOnDestroy([this](CHyprlandInputCaptureManagerV1* p) { this->onManagerResourceDestroy(p->resource()); });
```
no this->

vaxerski commented on src/protocols/InputCapture.cpp:
```diff
@@ -0,0 +1,75 @@
+
+    RESOURCE->setOnDestroy([this](CHyprlandInputCaptureManagerV1* p) { this->onManagerResourceDestroy(p->resource()); });
+
+    RESOURCE->setCapture([this](CHyprlandInputCaptureManagerV1* p) { this->onCapture(p); });
```
why? the functions are small enough, just plop them into this lambda

vaxerski commented on src/protocols/InputCapture.cpp:
```diff
@@ -0,0 +1,75 @@
+}
+
+void CInputCaptureProtocol::sendMotion(const Vector2D& absolutePosition, const Vector2D& delta) {
+    for (const UP<CHyprlandInputCaptureManagerV1>& manager : m_vManagers) {
```
const auto const&

3l0w replied:
Huh Clangd don't like `const auto const&`: duplicate 'const' declaration specifier. Its fine with just `const auto&`

vaxerski replied:
fair enough, odd tho

littleblack111 commented:
just a general question. will this be released in the next release or will it be taking a while, considering it's marked as draft

3l0w commented:
> just a general question. will this be released in the next release or will it be taking a while, considering it's marked as draft

Idk when i will be finished. if you need it, build it yourself

littleblack111 commented:
With your fork, will kvm software such as input leap work out of the box?

3l0w commented:
Yes and i am it to test

littleblack111 commented:
wow it was really troublesome compiling xdg-desktop-portal-hyprland. have to copy over the protocol to my /usr/share/hyprland-protocols(or clone it there), or i just removed the if line and forced it to get it from the repo.

but, it seems to just fail in input-leap:
```
[2024-10-09T19:43:32] ERROR: Failed to initialize InputCapture session, quitting: CreateSession failed
```
it was giving a error that it cant find the input-leap location before installing your hyprland & xdg-desktop-portal-hyprland fork. but that error is gone. but still failed.

3l0w commented:
Its strange that you have to copy the protocol file over, did you clone the submodules ? (`git submodule init && git submodule update`)
Also i forget to mention that you need to copy the file `hyprland.portal` over to a location that the xdg-desktop-portal will pick it up (on my arch install its over `/usr/share/xdg-desktop-portal/portals/hyprland.portal`) and dont forge to restart the xdg-desktop-portal after that.
Without that xdg-desktop-portal will not pickup xdph when forwarding InputCapture requests.

littleblack111 commented:
> Its strange that you have to copy the protocol file over, did you clone the submodules ? (git submodule init && git submodule update)

its the cmakelist file, it tried to grab from /usr/share/hyprland-protocols if exist, which doesn't have your protocol as it's not merged in the hyprland-protocol repo, thus i made it only grab from the gh repo.

i see, ill try that

3l0w commented:
Thanks for the feedback tho i didn't encounter most of the issues you describe, but i will try to investigate it. 
I am currently implementing a shortcut to force release the pointer and give it back to hyprland to prevent softlock.
The portal that i am implementing is not responsible for sharing the clipboard so there is really nothing i can do to make it work, it is a inputleap related issue.
I will try to implement something to make the cursor disappear, and also i need handle the cursor position send by the client when the capture is released maybe it will fix 4.
I am not yet sending modifier keys, i will impl it soon to see if it will fix 1 for you.

littleblack111 commented:
alright. ty! Appreciate it!

3l0w commented:
The latest commit added a new shortcut: `releaseinputcapture`
This shortcut force release the capturing of inputs in case of softlock.

Personally, I bind it like this: 
```
bind = $mainMod SHIFT, escape, releaseinputcapture
```

littleblack111 commented:
Nice! I'm currently on vacation without my laptop, so I'll try it when I get back

littleblack111 commented:
> A suggestion is that, disable this when mouse or keyboard is being capture/binded by programs(e.g. games that uses mouse to move camera)

if thats not possible then maybe add a dispatcher so user can script it to do so

littleblack111 commented:
https://github.com/user-attachments/assets/4f101b28-a21b-42fc-9850-8fb5477f5fea

3l0w commented:
> > A suggestion is that, disable this when mouse or keyboard is being capture/binded by programs(e.g. games that uses mouse to move camera)
> 
> if thats not possible then maybe add a dispatcher so user can script it to do so

You mean if a 3d game has kinda the mouse captured and move it to a border the input is captured and should not ?
I think its kinda a major issue if its the case, i will look into it

3l0w commented:
> trim.44B51EBB-2916-40A4-BFBB-7EBA4AF42874.MOV

If you have any steps to reproduce it it will be great

littleblack111 commented:
> > > A suggestion is that, disable this when mouse or keyboard is being capture/binded by programs(e.g. games that uses mouse to move camera)
> > 
> > 
> > if thats not possible then maybe add a dispatcher so user can script it to do so
> 
> You mean if a 3d game has kinda the mouse captured and move it to a border the input is captured and should not ? I think its kinda a major issue if its the case, i will look into it

yes, for example my macbook is left, and i moved my cursor in game(that i captured) to left, it will eventually go to my mac screen

3l0w commented:
It is quite strange that you where able to still move the mouse in your primary computer.

littleblack111 commented:
not sure, it just happend randomly, but before it happend, i was impl a new feature and was opening a nested Hyprland, but i wont try again today cuz it's pretty late and ill sleep

littleblack111 commented:
> > > A suggestion is that, disable this when mouse or keyboard is being capture/binded by programs(e.g. games that uses mouse to move camera)
> > 
> > 
> > if thats not possible then maybe add a dispatcher so user can script it to do so
> 
> You mean if a 3d game has kinda the mouse captured and move it to a border the input is captured and should not ? I think its kinda a major issue if its the case, i will look into it

related: https://www.reddit.com/r/kde/comments/u34lg0/wayland_gaming_and_mouse_capture/

littleblack111 commented:
a window rule to prevent inputcapture will be great too

3l0w commented:
If you make proposition like don't hesitate to explain it a little bit more, like what is supposed to do precisely. It will make my life simpler.
Also i've work on that since a few weeks, i will get back to it in the nexts weeks

littleblack111 commented:
oh ya sure.
so when you go in something like minecraft or basically any 3d games that captures you mouse. swing your mouse to left rapidly and it will go to the other screen.

for the windowrule one, just like other window rule, you want it prevent a specific window to trigger it, like `windowrulev2 = noinputcapture, title:Minecraft`

littleblack111 commented:
would love to see some work done, also numlock seem to be inverted to the client

littleblack111 commented:
please fix merge issues with upstream, theres a lot of changes...

3l0w commented:
> would love to see some work done, also numlock seem to be inverted to the client

I added smth do the send the modifers to the clients, but this has some drawback most notably num lock light wont blink if you pressed while input is captured, but it will work on the client as expected

3l0w commented:
> oh ya sure. so when you go in something like minecraft or basically any 3d games that captures you mouse. swing your mouse to left rapidly and it will go to the other screen.
> 
> for the windowrule one, just like other window rule, you want it prevent a specific window to trigger it, like `windowrulev2 = noinputcapture, title:Minecraft`

Idk if i need this window rule, because i think hyprland has mecanism to prevent mouse passing from one screen to another while in game

3l0w commented:
I added it, feel free to test

littleblack111 commented:
nice! theres still another small issue besides the cursor delay when reappear. I'm not sure what caused this, but `isLocked` seem to be kinda slow? for example, if the cursor is locked, then i immediately go the direction of the other device rapidly, it will still be going to the other device, and it will be glitched in a weird state - all the keyboard events and clicks work(on the other device), but i cannot move the mouse(prob cuz now the `isLocked` is true?), thus i cannot go back, so i have to use the releaseinputcapture dispatcher. however, it does work (no rapid movement for like 1-2 sec then it won't go to that side anymore)

littleblack111 commented:
> > would love to see some work done, also numlock seem to be inverted to the client
> 
> I added smth do the send the modifers to the clients, but this has some drawback most notably num lock light wont blink if you pressed while input is captured, but it will work on the client as expected

hmm, idk, the num lock is weird, it will not be lighted for first time you swap to the other device, but will stil be lighted later... idk it's very glitchy lol

littleblack111 commented:
additionally, the problem of modifiers being held down still presist

3l0w commented:
> additionally, the problem of modifiers being held down still presist

How do you know that they are held down ? What are exacly the symptomes that you observe ? 
I need a way to reproduce it, i have no idea why and how.

littleblack111 commented:
the symptoms is just when i type, it doensn't type and the system shortcuts keep getting triggerd, i figure if it's super(cmd) by pressing only tab etc(i solve this by just opening input leap, stop the server and restarting it). i have no idea how to reproduce it either, it just happens like 3-5times a day. not to mention the issue of it just disconnecting for no reason, no power state change(sleep etc) just worked on machine for a while(3+hr) then wen i try to go back to my mac, it just doesn't work(i solve this by just killall input-leap input-leaps && systemctl --user restart xdg-desktop-portal-hyprland && input-leap &)

don't get me wrong, this is still amazing work! thanks

littleblack111 commented:
the most annoying part rn is prob just needing to restart this like 20-50 times every day, unsure why but it will sometimes just randomly stop working(`killall input-leap input-leaps && sysu restart xdg-desktop-portal-hyprland.service; input-leap &` would fix it)

3l0w commented:
Huh, i never got a bug like that.
What are the symptoms? You send your mouse to the side and nothing happen ?
I think i need to add some debug logs to that so you can try to give me more info on that issue.
I will try to do that fairly quickly.

littleblack111 commented:
yep, just tries to go to macbook, all the way to the side, continue going, just stuck there and mouse still at the side.

not sure how would i collect debug logs as it'd take quite and while to realize(since obv im not gonna be using my macbook all the time), hence the log might just mostly be spam especially if its debug

littleblack111 commented on src/render/Renderer.cpp:
```diff
@@ -2290,7 +2291,9 @@ void CHyprRenderer::ensureCursorRenderingMode() {
+
+    const bool HIDE = m_sCursorHiddenConditions.hiddenOnTimeout || m_sCursorHiddenConditions.hiddenOnTouch || m_sCursorHiddenConditions.hiddenOnKeyboard ||
+            PROTO::inputCapture->isCaptured();
 
```
consider reschedule frame to immediately update like this:
```
g_pCompositor->scheduleFrameForMonitor(pMonitor)
```

3l0w replied:
Oh thanks, i did a little search for a method like this and did not found it

littleblack111 replied:
Aight, also look at my PR

littleblack111 reviewed (commented):
mb, please just resolve this

3l0w commented:
Yeah debug logs will be spamming things, but for now you can send me the logs of input-leap & xdph maybe it could help

littleblack111 commented:
Alright, input leap log in debug too?

3l0w commented:
Yeah maybe not

littleblack111 commented:
hmm it just randomly stopped working
xdph err:
```
 Input captured  EIS: 16:08:50 | ERROR | 🪲  Bug: eis_device_stop_emulating: missing call to eis_device_frame()
```

3l0w commented:
Okay i will commit a fix and a rebase soon

littleblack111 commented:
ok mb it's my patch's issue

3l0w commented:
Nah mb too, i need to send a frame after stop emulating i think

littleblack111 commented:
ok mb it's my patch's issue

> > trim.44B51EBB-2916-40A4-BFBB-7EBA4AF42874.MOV
> 
> If you have any steps to reproduce it it will be great

multiple input-leap instances. maybe we shall only accpet one as... im not sure what would be the use case of multiple instance of any kvm software requetsing multiple instnaces of the protocol

3l0w commented:
> > > trim.44B51EBB-2916-40A4-BFBB-7EBA4AF42874.MOV
> > 
> > 
> > If you have any steps to reproduce it it will be great
> 
> multiple input-leap instances. maybe we shall only accpet one as... im not sure what would be the use case of multiple instance of any kvm software requetsing multiple instnaces of the protocol

Nothing is said about that in the specs, maybe i will limit to one app being in the enable state

3l0w commented:
I can't understand in your logs when it started to work again. Do you have smth like the time when i doesn't work anymore ?

littleblack111 commented:
hmm... all of a sudden now the keybinds will trigger on the host even if mouse is captured on the client...

3l0w commented:
I rebased maybe i broke something

3l0w commented:
Okay i found why, i am pushing a fix

littleblack111 commented:
some more logs for wen it stopped working(nothin out of the ordinary from input-leap)

xdph:
[log.txt](https://github.com/user-attachments/files/18425854/log.txt)

littleblack111 commented:
same condition:

[log.txt](https://github.com/user-attachments/files/18468380/log.txt)

littleblack111 commented:
[log.txt](https://github.com/user-attachments/files/18545569/log.txt)

littleblack111 commented:
pretty sure its that last line cuz this time it worked like 10sec ago and now it doesnt

3l0w commented:
I see something called "transient" in the logs, what is it ?

3l0w commented:
Maybe i don't have the latest version of input-leap :thinking:

3l0w commented:
> pretty sure its that last line cuz this time it worked like 10sec ago and now it doesnt

I think i found why you got that last line but i am not sure,
Also can you use this command to get the logs: 
```bash
journalctl --user -u xdg-desktop-portal-hyprland -b
```

littleblack111 commented:
[patch.txt](https://github.com/user-attachments/files/18550464/patch.txt)
plz fix

3l0w commented:
> [patch.txt](https://github.com/user-attachments/files/18550464/patch.txt) plz fix

Bruh i did but forget to commit :cry:

littleblack111 commented:
[log.txt](https://github.com/user-attachments/files/18550535/log.txt)
more

3l0w commented:
> [log.txt](https://github.com/user-attachments/files/18550535/log.txt) more

The send of inputs stop everytime you got "file descriptor expected, object (10), message keymap(uhu)" ?

3l0w commented:
Which keymap do you use ? i think its related

littleblack111 commented:
> Which keymap do you use ? i think its related

default. i literally did not write a section in my hyprland conf for it.

but i use fcitx5

littleblack111 commented:
> I see something called "transient" in the logs, what is it ?

i see that too, not idea. not in my processes either

isaacgale commented on src/protocols/InputCapture.cpp:
```diff
@@ -0,0 +1,113 @@
+    hyprlandInputCaptureManagerV1KeymapFormat format;
+    int                                       fd;
+    uint32_t                                  size;
+    if (keyboard) {
```
Are these if/else conditions needed here after the one above? The code should never get down here unless it's a 'keyboard'. So the else could be deleted and also the if at the bottom of this method too. If I am wrong my apologies.

3l0w replied:
Yeah probably, i didn't quite write this code i stole it from another part of hyprland.
But it seems that the code where i have copied from has changed so i will update this accordingly

littleblack111 commented:
it seems like `hyprctl reload` would make it not work until a `killall input-leap input-leaps && sysu restart xdg-desktop-portal-hyprland.service; input-leap &`

3l0w commented:
Reminder to myself that i need to do some impl around #9930 to add some level of security

samvv commented:
Could it be that we now have https://wayland.app/protocols/wlr-virtual-pointer-unstable-v1 to do the same? Sorry if I'm wrong, I'm just now familiarizing myself with the protocol.

3l0w commented:
Yup its the reverse but someone based on my PR use it to impl the Remote Desktop Portal (hyprwm/xdg-desktop-portal-hyprland#308)
But i will use it at some point because it need to impl cursor teleportation when i release it.

3l0w commented:
> Anything I can do to help move this forward? I'm planning to make use of the Remote Desktop Portal, which relies on this one apparently.

Sure ive made the list of remaining tasks here: https://github.com/hyprwm/xdg-desktop-portal-hyprland/pull/268#issuecomment-2827672949
Feel free to make PRs to my forks

littleblack111 commented:
Please rebase main,
All the var names are wrong since https://github.com/hyprwm/Hyprland/commit/241a4935a244f403fa7108259075b04c81ed258f and similar commits

lxe commented:
I was able to get this along with https://github.com/hyprwm/xdg-desktop-portal-hyprland/pull/268 working. I think this needs a small change like https://github.com/lxe/Hyprland/commit/16adaa379b678d51547ccdab10588ba5c3e8368f and it's good to go... at least synergy/deskflow server works now.

3l0w commented:
> Gentle bump on this one
> 
Sorry i am on vacation for another week, i will rebase everything when i can and continue to work on it.
I will update the checkboxes on the pr desc when i will make progress.

lxe commented:
> > Gentle bump on this one
> 
> Sorry i am on vacation for another week, i will rebase everything when i can and continue to work on it. I will update the checkboxes on the pr desc when i will make progress.

You're all good! Thanks for your hard work. I've been rebasing and testing your implementation of the portal, protocol, and Hyprland and it feels stable. At least it's making my deskflow/barrier/synergy work well. 

The only bug was key repeat transfer but we can cross that bridge later.

lxe commented:
I've been rebasing this one and fixing conflicts here: https://github.com/lxe/Hyprland

3l0w commented:
whoops i noticed this, fixed it, but forgot to commit & push sorry

alerque commented:
Is there anything a user/tester/distro packager can do to help this PR along? Not having this feature across my distros is really starting to make Hyprland be a drag given my current machine layout. Is there an outline of things that still need to be addressed? Would it be helpful if I split out the bug fixes unrelated to the feature addition and submitted those in a separate PR to get reviews rolling in bit size chunks?

3l0w commented:
Testing and provide feedback is probably the best thing to do for now, the todo list is in the PR description.
Its really not ready to be deployed, there security issue that i am currently addressing. The PR is almost done, hold still !

lexxxel commented:
like I [commented](https://github.com/hyprwm/xdg-desktop-portal-hyprland/pull/268#issuecomment-3274893889) in the portel PR, input share with macOS from a gentoo host works as expected :+1:

lxe commented:
It's all good I've been rebasing and merging all your PRs and daily-driving latest hyprland, portal, and protocols with zero issues.

3l0w commented:
Ive pushed a huge refacto, feel free to test it

3l0w commented:
I think, this MR is almost done.
There is still features that i want to add (like shortcut who work when input captured), but this will be for a later mr

stauersbol commented:
I just updated my side, and I don't get the permission prompt at all?
```
[2025-09-29T08:46:30] INFO: starting core server process (desktop mode)
[2025-09-29T08:46:30] INFO: core config file: /home/stauersbol/.config/Deskflow/deskflow-server.conf
[2025-09-29T08:46:30] INFO: running command: /usr/bin/deskflow-core server -f --debug FATAL --name solaris --prevent-sleep --address :24800 -c /home/stauersbol/.config/Deskflow/deskflow-server.conf
[2025-09-29T08:46:30] IPC: started server, waiting for clients
[2025-09-29T08:46:31] IPC: client "Lukas-THINKPAD" has connected
```

Deskflow logs say I have a connection, but when I move my cursor to the right to where my other client is, then it doesn't appear?

3l0w commented:
Maybe, your protocols aren't up to date.... 
Check ur hyprland & xdph logs
For the permission thing maybe you haven't added this into your config:
```
ecosystem {
  enforce_permissions = 1
}
```

3l0w commented:
@vaxerski PR is ready for review

vaxerski commented on src/helpers/Eis.cpp:
```diff
@@ -0,0 +1,315 @@
+#include "Eis.hpp"
```
Wrong region, I'd put it under managers/input.

vaxerski commented on src/helpers/Eis.hpp:
```diff
@@ -0,0 +1,59 @@
+    int         getFileDescriptor();
+
+  private:
+    bool stop   = false;
```
style: members are `m_camelCase`

vaxerski commented on src/managers/KeybindManager.cpp:
```diff
@@ -782,6 +785,14 @@ SDispatchResult CKeybindManager::handleKeybinds(const uint32_t modmask, const SP
 
             m_passPressed = sc<int>(pressed);
 
+            // We only process the releaseinputcapture dispatcher when input capture is active
```
do this inside ::releaseInputCapture. It's ugly and cluttered here. Just return an error when it's not captured.

3l0w replied:
Ive simplified the if, tell me if its good for u

vaxerski commented on src/protocols/InputCapture.cpp:
```diff
@@ -0,0 +1,410 @@
+}
+
+void CInputCaptureProtocol::destroyResource(CInputCaptureResource* resource) {
+    Debug::log(LOG, "Destroying resource {}", resource->sessionId);
```
don't log this, wayland_debug would already do so.

vaxerski commented on src/protocols/InputCapture.cpp:
```diff
@@ -0,0 +1,410 @@
+
+void CInputCaptureProtocol::destroyResource(CInputCaptureResource* resource) {
+    Debug::log(LOG, "Destroying resource {}", resource->sessionId);
+    clearBarriers(resource->sessionId);
```
this would be better off in the dtor of the resource (bugprone)

vaxerski commented on src/protocols/InputCapture.cpp:
```diff
@@ -0,0 +1,410 @@
+}
+
+void CInputCaptureProtocol::updateKeymap() {
+    for (auto& session : m_Sessions)
```
{} around for

vaxerski commented on src/protocols/InputCapture.cpp:
```diff
@@ -0,0 +1,410 @@
+
+std::optional<SP<CInputCaptureResource>> CInputCaptureProtocol::getSession(std::string sessionId) {
+    auto result = std::ranges::find_if(m_Sessions, [sessionId](SP<CInputCaptureResource> elem) { return sessionId == elem->sessionId; });
+    if (result == m_Sessions.end()) {
```
no {} around if

vaxerski commented on src/protocols/InputCapture.hpp:
```diff
@@ -0,0 +1,109 @@
+#include <wayland-server.h>
+
+enum eClientStatus : uint8_t {
+    CREATED,   //Is ready to be activated
```
enums need a prefix to avoid clashes. `CLIENT_STATUS_CREATED`, for example.

vaxerski commented on src/protocols/InputCapture.hpp:
```diff
@@ -0,0 +1,109 @@
+struct SBarrier {
+    std::string sessionId;
+    uint        id;
+    int         x1, y1, x2, y2;
```
no uninitialized primitives (bugprone)

vaxerski commented on src/protocols/InputCapture.hpp:
```diff
@@ -0,0 +1,109 @@
+	UP<CEmulatedInputServer> eis;
+
+    //
+    uint32_t             capabilities = 0;
```
members are m_camelCase

vaxerski commented on src/protocols/InputCapture.hpp:
```diff
@@ -0,0 +1,109 @@
+    SP<HOOK_CALLBACK_FN> monitorCallback;
+
+    //
+    void deactivate();
```
functions above vars in general (style)

fn
fn
fn

var
var
var

vaxerski reviewed (commented):
cant comment on the tracy submodule bump: pls remov

overall nice work!

lexxxel commented on src/helpers/Eis.hpp:
Please pardon my Intrusion! Why do the files for CEmulatedInputServer are named Eis and not EmulatedInputServer?
I got triggered because Eis is german for ice and ICE is not that popular in the US right now, right? Maybe just dodge that `bullet` ....

alerque replied:
You can argue a more verbose name would be better than terse acronyms for clarity's sake, but loose the flimsy political pandering excuse. Even in one language if something technical (say Integrated Control Elevator) happened to abbreviate to ICE and that made it into a file name it would be neither here nor there. Objecting to that same thing through an extra level of abstraction via another language is just silliness.

3l0w replied:
As much i as hate the ICE, EIS is the technical short name of [libei server](https://libinput.pages.freedesktop.org/libei/api/group__libeis.html). Maybe i will rename the class to EIS, because *in my opinion*, EIS give more context to what it is than EmulatedInputServer

3l0w commented:
Nix build seems to be borken because hyprland protocols is not up to date idk why

3l0w commented:
> <img alt="20251007_12h42m51s" width="362" height="148" src="https://private-user-images.githubusercontent.com/43317083/498275860-7dae3597-f356-4d6c-a79b-bd6c0f58fe51.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk5MjgyNjIsIm5iZiI6MTc1OTkyNzk2MiwicGF0aCI6Ii80MzMxNzA4My80OTgyNzU4NjAtN2RhZTM1OTctZjM1Ni00ZDZjLWE3OWItYmQ2YzBmNThmZTUxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA4VDEyNTI0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY3NDhjMGZiYTI0OGNkNDdlZWMxOWMwNTQxNDA1ZTUwNDhlNTFiYjA1NDFiODZjN2JhYjFjZjg5OTFhZGNmNjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YaSiuXw3ZMJ391oRPNGXPOlhmrWISf85mc-1zgqDOUQ"> this is still here

My bad, ive done a fixup every commit and removed the tracy bump

vaxerski reviewed (commented):
lgtm other than style.

Would it be feasible to write some tests for this? I don't know what the libei space looks like and if it's easy to write a simple client to send some requests

vaxerski commented:
would be great, thanks.

gac3k commented:
@3l0w apologize for pinging again, but do you need some help here? Can we take-over some bits anyhow just to speed things up a bit? :)

vaxerski commented:
you seem to have diverged the branches weirdly

3l0w commented:
> you seem to have diverged the branches weirdly

Yup ive dropped https://github.com/hyprwm/Hyprland/commit/1912ba87e9a491e72c8ef735dcbaf473c6e1ba93 because it prevented to build hyprland without hyprpaper-git, i was waiting the breaking changes to be released in a hyprpaper release. A simple rebase against main should fix

acheong08 commented:
> and? did anyone test?

I've been using this for a while (updated to latest ~2 days ago). Generally works pretty well. Ran into one issue that might be a deskflow bug where when the cursor goes off screen when the client isn't connected, it gets stuck and can't return. Have to restart to get the cursor back.

Kaysera commented:
> Ran into one issue that might be a deskflow bug where when the cursor goes off screen when the client isn't connected, it gets stuck and can't return.

Same bug here. And sometimes it is not able to properly detect the boundaries when I connect a new screen and I have to restart the xdg portal to get it running (using `kill`). Aside from that, after a restart it works really well.

3l0w commented:
> > Same bug here. And sometimes it is not able to properly detect the boundaries when I connect a new screen and I have to restart the xdg portal to get it running (using `kill`). Aside from that, after a restart it works really well.
> 
> How do you restart the xdg portal while input is captured? I can't type or anything

You can have a bind to release the inputs
```
bind = $mainMod SHIFT, escape, releaseinputcapture
```

Strange issue with deskflow tho, but i believe the impl kinda sucks beaucoup they don't release the zones when not using it, and input-leap always reserve every possible zones
And also strange to have to restart xdph for the screen to work....

ZeroSegFault committed:
fix: repeat captured keyboard input & harden EIS setup

extra-salad commented:
I tested this with Deskflow 1.26.0 using:

- Hyprland from this InputCapture branch
- xdg-desktop-portal-hyprland from the matching `feat/input-capture-impl` branch
- `xdg-desktop-portal` 1.20.4
- `libei` 1.5.0

InputCapture is exposed correctly and Deskflow basically works, but I found a reproducible issue with vertically offset display outputs.

Working monitor layout:

```conf
monitor = DP-2, 2560x1440@165.002, 0x0, 1
monitor = HDMI-A-1, 1920x1080@60.000, 2560x0, 1
```

Broken layout:

```conf
monitor = DP-2, 2560x1440@165.002, 0x0, 1
monitor = HDMI-A-1, 1920x1080@60.000, 2560x180, 1
```

With the offset layout, Deskflow still detects/captures input, however, the cursor gets trapped at the internal border between the two local monitors. I assumes it's trapped in a way where it keeps switching from one machine to another as there is a barrier created in-between.

In this layout the real shared seam is only:

```text
x = 2560
y = 180..1259
```

but the left monitor is taller:

```text
DP-2:      x=0..2559     y=0..1439
HDMI-A-1: x=2560..4479  y=180..1259
```

When I remove the vertical offset and align both outputs at y=0, the trapping goes away.

Deskflow logs also reveal 4 failed barrier for the working layout:

```text
[2026-05-06T00:09:14.596] WARNING: failed to apply barrier 1 (2560/0-2560/1079)
[2026-05-06T00:09:14.596] WARNING: failed to apply barrier 2 (4480/0-4480/1079)
[2026-05-06T00:09:14.596] WARNING: failed to apply barrier 1 (0/0-0/1439)
[2026-05-06T00:09:14.596] WARNING: failed to apply barrier 2 (2560/0-2560/1439)
```

While broken layout only gets 2 failed barriers:

```text
[2026-05-06T00:11:41.140] WARNING: failed to apply barrier 1 (2560/180-2560/1259)
[2026-05-06T00:11:41.140] WARNING: failed to apply barrier 1 (0/0-0/1439)
```

prurigro commented:
@extra-salad out of curiosity, does it work for you if you change the scale factor for the higher res monitor? It's broken for me, but I have my lower res monitor on the left side.

extra-salad commented:
@prurigro 
> @extra-salad out of curiosity, does it work for you if you change the scale factor for the higher res monitor? It's broken for me, but I have my lower res monitor on the left side.

Scaling seems to break InputCapture barrier handling. With DP-2 scaled to 1.25 and HDMI placed at the logical boundary:

```conf
monitor = DP-2, 2560x1440@165.002, 0x0, 1.25
monitor = HDMI-A-1, 1920x1080@60.000, 2048x0, 1
```

Deskflow fails creating any borders and reports with `failed to set up ei: Operation not permitted`, although that's probably just a side effect and XDPH/Hyprland logs are more specific:

```text
hyprland_input_capture_v1#...: error 1: The barrier id 1 is invalid
```

A scale intended to match the 1080p monitor height also fails:
```conf
monitor = DP-2, 2560x1440@165.002, 0x0, 1.333333
monitor = HDMI-A-1, 1920x1080@60.000, 1920x0, 1
```
with:
```text
hyprland_input_capture_v1#...: error 1: The barrier id 3 is invalid
```

Also fails with scales < 1.

Note: I'm pretty new on Linux, so I'm not sure if system reboots are needed for tests alike and I haven't done them when testing these cases, although I did run these in-between each test:

```bash
hyprctl reload
systemctl --user restart xdg-desktop-portal-hyprland xdg-desktop-portal
pkill -x deskflow-core
pkill -x deskflow
```

extra-salad commented:
Another nitpick that I've noticed and not sure if this is intentional or relevant, but it's not present on Plasma:

When cursor goes to another machine there's a slight delay before the cursor on the main machine disappears. I think this is purely visual and it seems like it does not affect the actual input capture on the target machine as the transitions otherwise look smooth.

https://github.com/user-attachments/assets/84bd8661-2875-4677-8eb7-6f410a0164ca

I also can see that in this implementation the cursor completely stops moving once it hits the remote machine, while on Plasma it never stops moving and only becomes invisible. Which does give style points for how transition between the machines look & feel, but on the other hand it is a bit annoying to see focus effects from a hidden moving cursor on the main machine while controlling the remote.

prurigro commented:
@extra-salad hmm, Deskflow should start and think it's running if it's like my tests- it just doesn't allow me to cross the edge of my display into the next one. I think I've run into an error like you're talking about though, after playing with screen settings (possibly specifically scale factors), and if I'm remembering correctly I was able to fix it by logging out and back in. I didn't check the hyprland logs to compare that end though, and I haven't hit it in quite a while so take with a grain of salt...

In theory a reboot wouldn't be necessary vs simply logging out and back in. That said, you'll want to make sure you have `KillUserProcesses=yes` and `UserStopDelaySec=0` in /etc/systemd/logind.conf and then logout with something like `loginctl kill-session "$XDG_SESSION_ID"` so you aren't leaving any user processes behind that are bound to the old session before logging back in.

Regarding your nitpick- funny, on gnome and sway the cursor just sits there forever until you come back :)

prurigro commented:
@extra-salad you can actually get key repeat working with ZeroSegFault's branch @ https://github.com/ZeroSegFault/Hyprland#branch=zerosegfault/input-capture-hyprland and https://github.com/ZeroSegFault/xdg-desktop-portal-hyprland#branch=input-capture-portal

prurigro commented:
@extra-salad Oooh, fixed scaling sounds good! I'll give it a test shortly :)

3l0w's is the correct place. Like you, ZeroSegFault made his changes on his own fork of that, and last I checked was creating a PR to get them merged into the upstream PR (you can see the conversation around that at the end of the conversation here: https://github.com/hyprwm/xdg-desktop-portal-hyprland/pull/268).

prurigro commented:
@extra-salad I believe gnome uses libei in their wayland implementation too... Interesting that spoofing didn't work. There must be something different about the values getting handed over when the display is 3840x2160 and reporting at 1920x1080 because of the scale factor vs telling the monitor to render 1920x1080 (which does work). Capture not getting handed over does seem like a possibility...

This is my monitor layout- the green ones are the server. Deskflow is configured so the barriers are the sides that are touching, and for testing I added a barrier at the top of the green displays that moves to the bottom of the yellow ones, and the right of the green displays that moves to the left of the red one. Everything works when the rightmost green one is set to 1920x1080, and it works at 3840x2160 provided the left one's y position (in hyprland) is against the bottom.  Setting it to 3840x2160 with a scale factor of 2 however, breaks the bottom barrier on both displays and the right one on the right display.

<img width="877" height="341" alt="layout" src="https://github.com/user-attachments/assets/eefcc131-a076-4e4b-90da-086e2f4492bf" />

Maybe your right display works because it isn't the one that's scaled?

I appreciate all the work you're putting into this btw!

extra-salad committed:
Feat/input capture impl (#10)

* input-capture: suppress local keybinds while captured

Captured keyboard events were still processed by the normal keybind path before the captured check suppressed delivery to local clients. This allowed compositor binds such as SUPER shortcuts to run locally while the same events were sent to the remote EIS client.

Run keybind handling in a capture-only mode while input capture is active, skipping internal shortcuts and allowing only the releaseinputcapture escape hatch.

* keybinds: add input capture exception flag

Add bindx / allow_input_capture for keybinds that should remain available while input capture is active. This lets users define an explicit local escape bind that can release capture and optionally perform additional local actions.

Expose the flag through Lua keybind options, hyprctl binds output, and Lua keybind metadata.

* input-capture: validate barriers against logical monitor size

Use the monitor logical size already computed by Hyprland when validating input-capture barriers instead of deriving it again from physical pixels and scale.

Recomputing logical size can disagree with xdg-output geometry for fractional scales, causing valid portal barriers to be rejected with invalid_barrier protocol errors.

vaxerski commented on src/config/values/ConfigValues.cpp:
```diff
@@ -710,6 +710,12 @@ std::vector<SP<IValue>> Values::getConfigValues() {
         MS<Bool>("experimental:wp_cm_1_2", "Allow wp-cm-v1 version 2", false),
 
+		/*
+		 * inputcapture: 
```
input_capture:

vaxerski commented on src/managers/input/Eis.hpp:
```diff
@@ -0,0 +1,62 @@
+
+    //
+
+    std::string socketPath;
```
still here

vaxerski commented on src/managers/input/Eis.hpp:
```diff
@@ -0,0 +1,62 @@
+#include <string>
+#include <wayland-server-core.h>
+
+struct SKeymap {
```
put this inside CEis to not pollute

vaxerski commented on src/protocols/InputCapture.hpp:
```diff
@@ -0,0 +1,122 @@
+#include <optional>
+#include <wayland-server.h>
+
+enum eClientStatus : uint8_t {
```
put inside CInputCaptureProtocol or Resource

vaxerski commented on src/protocols/InputCapture.hpp:
```diff
@@ -0,0 +1,122 @@
+    CLIENT_STATUS_STOPPED    //Can no longer be activated
+};
+
+struct SBarrier {
```
same

vaxerski commented on src/config/values/ConfigValues.cpp:
```diff
@@ -711,6 +711,12 @@ std::vector<SP<IValue>> Values::getConfigValues() {
+		 * input_capture: 
+		 */
+        MS<Bool>("inputcapture:capture_modifiers", "If enabled, modifiers are also captured and sent to the program", false),
+        MS<Bool>("inputcapture:enforce_barriers", "If enabled, throw a wayland error when a invalid barrier is received", true),
```
you didn't update these to input_capture.

vaxerski commented on src/managers/input/InputManager.cpp:
```diff
@@ -1659,8 +1683,15 @@ void CInputManager::onKeyboardMod(SP<IKeyboard> pKeyboard) {
 
-    auto       MODS = pKeyboard->m_modifiersState;
+    if (*PSENDMOD) {
+        //TODO: from config
```
still a todo or stale comment?

vaxerski reviewed (dismissed):
this is nice

vaxerski reviewed (approved):
it's so good to see
the world is alive
